[JENKINS-39232] Walk the DelegatingComputerLauncher instances when checking if JNLPComputerLauncher

stephenc opened the pull request:
See [JENKINS-39232](https://issues.jenkins-ci.org/browse/JENKINS-39232)

@jenkinsci/code-reviewers @reviewbybees

Replaces #2601

stephenc committed:
[JENKINS-39232] Walk the DelegatingComputerLauncher instances when checking if JNLPComputerLauncher

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

oleg-nenashev commented:
I do not believe it's a perfect fix since anybody can implement his own launcher using `ComputerLauncher` API. And there are many examples: https://github.com/search?q=org%3Ajenkinsci+%22extends+computerlauncher%22&type=Code

I agree it worths adding support of Delegating launchers, but it does not fix the issue

stephenc committed:
[JENKINS-39232] Add a system property to enable legacy plugins to continue to work until they get fixed

daniel-beck commented on core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java:
```diff
@@ -37,6 +41,16 @@
+     * {@link ComputerLauncher} when then should have extended {@link DelegatingComputerLauncher}
+     *
+     * @since FIXME
+     */
```
`@Extension(NoExternalUse.class)`

stephenc committed:
[JENKINS-39232] Be more forgiving of plugins

oleg-nenashev commented on core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java:
```diff
@@ -37,20 +44,60 @@
+            Method getDelegate = launcher.getClass().getMethod("getDelegate");
+            return ComputerLauncher.class.isAssignableFrom(getDelegate.getReturnType()) ? (ComputerLauncher) getDelegate
+                    .invoke(launcher) : null;
+        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
```
ReflectiveOperationException?

oleg-nenashev commented on core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java:
```diff
@@ -37,20 +44,60 @@
+                launcher = ((DelegatingComputerLauncher) launcher).getLauncher();
+            } else if (launcher instanceof ComputerLauncherFilter) {
+                launcher = ((ComputerLauncherFilter) launcher).getCore();
+            } else if (null != (l = getDelegate(launcher))) {  // TODO remove when all plugins are fixed
```
Does not guarantee a success for old plugins from what I see

oleg-nenashev commented on core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java:
```diff
@@ -37,20 +44,60 @@
+            } else {
+                if (disableStrictVerification) {
+                    LOGGER.log(Level.WARNING, "Connecting {0} as a JNLP agent where the launcher does not mark itself "
+                            + "correctly as being a JNLP agent", clientName);
```
Please explicitly provide a class of the launcher

oleg-nenashev reviewed (changes requested):
If we want to go in such way, we need better diagnostic messages. :bug: for it.

I still have concerns about such approach due to impact on plugins, but it's a subject for voting by @jenkinsci/core

stephenc committed:
[JENKINS-39232] Yet more forgiveness of plugins but define the forward contract

stephenc committed:
[JENKINS-39232] Provide diagnostics when this issue is encountered

- Since the failure of strict verification is an issue, we should log it as a warning always

oleg-nenashev commented:
I approve the code change itself since it makes the situation better that it is in 2.27. So 🐝 and 👍 as a partial fix.

But I still do not feel it's a right approach to nuke so many instances. Hence as a community member I keep 👎 regarding the change as a **complete** fix of JENKINS-39232

oleg-nenashev commented:
@reviewbybees done
@jenkinsci/code-reviewers We really really really want to get it in the next weekly

oleg-nenashev commented:
P.S: I will follow-up on the default behavior

oleg-nenashev commented:
I will merge it into the next weekly if nobody shouts

KostyaSha commented on core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java:
```diff
@@ -37,20 +44,60 @@
         return computer != null;
     }
 
+    private static ComputerLauncher getDelegate(ComputerLauncher launcher) {
```
comment wasn't addressed?

KostyaSha commented on core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java:
```diff
@@ -37,20 +46,82 @@
+                return (ComputerLauncher) getDelegate.invoke(launcher);
+            }
+        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
+            // ignore
```
you can name `ignore` instead `e` to disable warnings static analysing tools.
